StateMachine/StateMachine.swift

```diff
@@ -0,0 +1,180 @@
+    the handler has run.
+    
+    - Note: Only one handler can be registered for an action.
+    
+    - Warning: Make sure to avoid retain loops in your code. For example, if you setup the machine in a variable of your
+    view controller and you then try to access that view controller from the actionHandler, you should use `[weak self]` 
+    or `[unowned self]` for the handler.
+
+    - parameter action: The action name
+    - parameter fromStates: One or more states from which the action can be performed, nil if any state is acceptable
+    - parameter toStates: The states that the action handler may return, nil if any state is acceptable
+    - parameter actionHandler: The handler to run when performing the action
+    */
+    public func registerAction(action: A, fromStates: Set<S>?, toStates: Set<S>?, actionHandler: ActionHandler) {
+        actions[action] = (fromStates, toStates, actionHandler)
+    }
+
+    /** Performs a registered action
+
+    The action will only be performed if the machine is in one of the states for which the action was registered.
+
+    - parameter action: The action to perform
+    - returns: Returns the new state if the action was run, or nil if the action not run
+    */
+    public func performAction(action: A) -> S? {
+        if let (fromStates, toStates, actionHandler) = actions[action] {
+            if (fromStates == nil || fromStates?.contains(state) == true) {
+                let newState = actionHandler(machine: self)
+                if (toStates == nil || toStates?.contains(newState) == true) {
+                    state = newState
+                    return state
+                }
+                print("WARNING: The action handler for \"\(action)\" returned the state \"\(newState)\" but the state machine expects one of these states: \(toStates!). State kept at \"\(state)\".")
+                return nil
+            }
+        }
+        return nil
+    }
+
+    private var changes: [(Set<S>?, Set<S>?, ChangeHandler)] = []
+
+    /** Registers a handler to run on state change
+
+    Registers a handler to run when the state of the machine changes. A change handler only gets run if the old state
+    occurs in the fromStates set and the new state occurs in the toStates set. A nil set for either means any state 
+    is acceptable.
+    
+    - Note: If an action set the same state as was already set, these handlers get run too.
+    - Note: The change handlers will be run in the order they were registered.
+
+    - parameter fromStates: The handler is only run if the old state is in this set. If nil, any state is acceptable.
+    - parameter toStates: The handler is only run if the new state is in this set. If nil, any state is acceptable.
+    - parameter changeHandler: The handler to run
+    */
+    public func onChange(fromStates fromStates: Set<S>? = nil, toStates: Set<S>? = nil, changeHandler: ChangeHandler) {
```

Comment:
Swift compiler warning: 'fromStates fromStates' can be expressed more succinctly as '#fromStates'